servers/visual_server.cpp

```diff
@@ -467,29 +502,44 @@ Error VisualServer::_surface_set_data(Array p_arrays, uint32_t p_format, uint32_
 
 				PoolVector<real_t>::Read read = array.read();
 				const real_t *src = read.ptr();
-
 				if (p_format & ARRAY_COMPRESS_TANGENT) {
+					const float bias = 1.0f / 127;
 					for (int i = 0; i < p_vertex_array_len; i++) {
-						int8_t xyzw[4] = {
-							(int8_t)CLAMP(src[i * 4 + 0] * 127, -128, 127),
-							(int8_t)CLAMP(src[i * 4 + 1] * 127, -128, 127),
-							(int8_t)CLAMP(src[i * 4 + 2] * 127, -128, 127),
-							(int8_t)CLAMP(src[i * 4 + 3] * 127, -128, 127)
+						Vector3 source(src[i * 4 + 0], src[i * 4 + 1], src[i * 4 + 2]);
+						Vector2 res = norm_to_oct(source);
+
+						res.y = res.y * 0.5f + 0.5;
```

Comment:
nitpick, but either use the "f" for both or neither. It is a little awkward to have 0.5f and 0.5 next to each other